yarn-project/acir-simulator/src/client/simulator.ts

```diff
@@ -161,28 +161,24 @@ export class AcirSimulator {
     storageSlot: Fr,
     notePreimage: Fr[],
   ) {
-    let artifact: FunctionArtifactWithDebugMetadata | undefined = undefined;
-
-    // Brute force
-    for (let i = notePreimage.length; i < MAX_NOTE_FIELDS_LENGTH; i++) {
-      const signature = `compute_note_hash_and_nullifier(Field,Field,Field,[Field;${i}])`;
-      const selector = FunctionSelector.fromSignature(signature);
-      try {
-        artifact = await this.db.getFunctionArtifact(contractAddress, selector);
-        if (artifact !== undefined) break;
-      } catch (e) {
-        // ignore
-      }
+    const artifact: FunctionArtifactWithDebugMetadata | undefined = await this.db.getFunctionArtifactByName(
+      contractAddress,
+      'compute_note_hash_and_nullifier',
+    );
+    if (!artifact) {
+      throw new Error(
+        `Mandatory implementation of "compute_note_hash_and_nullifier" missing in noir contract ${contractAddress.toString()}.`,
+      );
     }
 
-    if (artifact == undefined) {
+    const maxNoteFields = (artifact.parameters[artifact.parameters.length - 1].type as ArrayType).length;
+    if (maxNoteFields < notePreimage.length) {
       throw new Error(
-        `Mandatory implementation of "compute_note_hash_and_nullifier" missing in noir contract ${contractAddress.toString()}.`,
+        `The note being processed has ${notePreimage.length} fields, while "compute_note_hash_and_nullifier" can only handle a maximum of ${maxNoteFields} fields. Please consider increasing the allowed field size to accommodate all notes generated from the contract.`,
```

Comment:
this error message seems different to what you used in the test file ?

Reply:
The test matches part of the message, which (I think) is the most important part that tells users the "compute_note_hash_and_nullifier" function current only takes n fields.

Reply:
ah yea!